sled-agent/src/updates.rs

```diff
@@ -127,7 +127,7 @@ impl UpdateManager {
 
                 let response = nexus
                     .cpapi_artifact_download(
-                        nexus_client::types::KnownArtifactKind::ControlPlane,
+                        &KnownArtifactKind::ControlPlane.to_string(),
```

Comment:
There are two parts to this:

1. For the DB schema, I do think we must use arbitrary strings. In the case of an update, an old version of Nexus must be comfortable doing something reasonable with artifacts present in newer TUF repos. I don't really see a good way out of this.
2. For the HTTP endpoints--assuming that the DB schema must be this way, is there any path to providing something like KnownArtifactKind at all?

Reply:
I don't think I'm ready to concede that the DB schema has to use strings instead of enums. We use enums in the DB schema many places that may grow over time (metric producers, network interface kind, zone types, etc.), and we're going to have to be able to add variants to those. I don't know exactly how we're doing that yet, but I don't see why artifact kinds would be any different. If we require schema migrations to follow the same kind of rules as what we've talked about for API versioning, I assume it'd be something like "break into two steps: step one just updates the enum and must complete before step two uses it".

Reply:
(Sorry I'm just coming up to speed on this and I'm kind of thinking out loud!)

What would an old version of Nexus be able to sensibly do with an artifact kind that it doesn't know about?  It _could_ distribute it, but how would it know who to distribute it to or when?  Or is it that it can't do anything (yet), but just needs to store the metadata until the upgrade?

I assume the risks we're worried about (at least, the ones I'm worried about) are that using strings makes it super easy to (1) accidentally put an _actually_ invalid value into the field (like `control_plane_zone` instead of `control_plane`) and (2) write code that operates on input that it doesn't correctly support but doesn't realize that.  Using an enum is a nice way to represent for any hunk of code that it at least in principle understands what it's working with.

Would it help to store both a "raw" artifact kind as a string _and_ an enum that includes an "unknown" variant?  Use the enum where possible.  But after any upgrade, after the schema update, go through and update the enum value for anything that's "unknown" based on the raw value.  This is a little more work, but I think not much? and gives consumers of this information a way to use the well-typed thing.

Another approach would be to store just a string in the database, but have the model type provide a Rust enum of known values.  One option here would be to include an "Unknown" variant with the raw string value.  But maybe a better one is that we provide two model types for this table: one uses a Rust enum _without_ the "Unknown" variant and the other provides the string.  Hopefully most consumers would use the first one, but in the rare cases where we need to be able to talk about the data we don't understand, we can do that.

I'm not sure -- just brainstorming here.

Reply:
@jgallagher:

> I don't think I'm ready to concede that the DB schema has to use strings instead of enums. We use enums in the DB schema many places that may grow over time (metric producers, network interface kind, zone types, etc.), and we're going to have to be able to add variants to those. I don't know exactly how we're doing that yet, but I don't see why artifact kinds would be any different. If we require schema migrations to follow the same kind of rules as what we've talked about for API versioning, I assume it'd be something like "break into two steps: step one just updates the enum and must complete before step two uses it".

I think one major difference is unlike all the other examples, artifact kinds are part of the update process. In your proposed model, step one to update the enum must somehow happen before the artifacts are ingested. This seems like a circular problem.

There are probably ways to resolve it. For example, one thing that came to mind as I was writing this is that we could imagine shipping a "pre-ingest" schema update within the TUF repo -- and then the current version of Nexus is responsible for applying that schema update before processing the TUF repo. But that seems like it blows up the complexity of the process quite significantly, for a benefit that doesn't (at first) appear commensurate.

@davepacheco:

> What would an old version of Nexus be able to sensibly do with an artifact kind that it doesn't know about? It could distribute it, but how would it know who to distribute it to or when? Or is it that it can't do anything (yet), but just needs to store the metadata until the upgrade?

I think at the very least the old version of Nexus should store the artifact, so that if the new version of Nexus can use it somehow it'll know what to do with it.

One of the main consumers of artifacts is sled-agent -- a possible state where a new version of sled-agent hits an old version of Nexus, asking for an artifact that it (but not the old Nexus) knows about.

> Would it help to store both a "raw" artifact kind as a string and an enum that includes an "unknown" variant? Use the enum where possible. But after any upgrade, after the schema update, go through and update the enum value for anything that's "unknown" based on the raw value. This is a little more work, but I think not much? and gives consumers of this information a way to use the well-typed thing.

Hmm, what would the benefit be of doing this in persistent storage, versus doing it in code? We already have [roughly this conversion](https://github.com/oxidecomputer/omicron/blob/823f90c0edd0407511372eb846976a157421cecb/common/src/update.rs#L164-L167) between `ArtifactKind` and `KnownArtifactKind`.

> But maybe a better one is that we provide two model types for this table: one uses a Rust enum without the "Unknown" variant and the other provides the string.

Right, this is roughly `ArtifactKind` vs `KnownArtifactKind`.

Reply:
> I think one major difference is unlike all the other examples, artifact kinds are part of the update process. In your proposed model, step one to update the enum must somehow happen before the artifacts are ingested. This seems like a circular problem.

We could require this to be split over two updates; the first update adds the new case(s) to the enum, and the second uses them. This seems similar to the plan for backwards-incompatible API changes, where the first update is still compatible with previous versions, and the second update is only compatible with the immediately-prior version (which brings the schema in line with what that second update expects to be there). But maybe this is more onerous than it's worth, _unless_ we end up needing to split updates like this anyway.

I'm left with a kind of uneasy feeling here - if the right thing to do is use strings, I'm worried there are other places where we're using serde (especially outside of API interfaces for which we have a plan) that are going to be more painful than we realize. But I don't really disagree with any of your arguments, and I don't want to hold up this PR on an uneasy feeling, so I'm happy to move forward with what you have.